insecure/corruptible/network.go

```diff
@@ -0,0 +1,370 @@
+// engine of the conduit is not going to use it anymore, so the channel can be closed safely.
+func (n *Network) EngineClosingChannel(channel channels.Channel) error {
+	return n.conduitFactory.UnregisterChannel(channel)
+}
+
+// eventToMessage converts the given application layer event to a protobuf message that is meant to be sent to the attacker.
+func (n *Network) eventToMessage(
+	event interface{},
+	channel channels.Channel,
+	protocol insecure.Protocol,
+	targetNum uint32,
+	targetIds ...flow.Identifier) (*insecure.Message, error) {
+
+	payload, err := n.codec.Encode(event)
+	if err != nil {
+		return nil, fmt.Errorf("could not encode event: %w", err)
+	}
+
+	myId := n.me.NodeID()
+	return &insecure.Message{
+		ChannelID: channel.String(),
+		OriginID:  myId[:],
+		TargetNum: targetNum,
+		TargetIDs: flow.IdsToBytes(targetIds),
+		Payload:   payload,
+		Protocol:  protocol,
+	}, nil
+}
+
+func (n *Network) generateExecutionReceipt(result *flow.ExecutionResult) (*flow.ExecutionReceipt, error) {
+	// TODO: fill spock secret with dictated spock data from attacker.
+	return ingestion.GenerateExecutionReceipt(n.me, n.receiptHasher, n.spockHasher, result, []*delta.SpockSnapshot{})
+}
+
+func (n *Network) generateResultApproval(attestation *flow.Attestation) (*flow.ResultApproval, error) {
+	// TODO: fill spock secret with dictated spock data from attacker.
+	return verifier.GenerateResultApproval(n.me, n.approvalHasher, n.spockHasher, attestation, []byte{})
+}
+
+// AttackerRegistered returns whether an attacker has registered on this corruptible network instance.
+func (n *Network) AttackerRegistered() bool {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	return n.attackerInboundStream != nil
+}
+
+// ConnectAttacker is a gRPC end-point for this corruptible network that lets an attacker register itself to it, so that the attacker can
+// control its ingress and egress traffic flow.
+// Registering an attacker on a networking layer is an exactly-once immutable operation,
+// any second attempt after a successful registration returns an error.
+func (n *Network) ConnectAttacker(_ *empty.Empty, stream insecure.CorruptibleConduitFactory_ConnectAttackerServer) error {
+	n.mu.Lock()
+	n.logger.Info().Msg("attacker registration called arrived")
+	if n.attackerInboundStream != nil {
+		n.mu.Unlock()
+		return fmt.Errorf("could not register a new attacker, one already exists")
+	}
+	n.attackerInboundStream = stream
+
+	n.mu.Unlock()
+	n.logger.Info().Msg("attacker registered successfully")
+
+	// WARNING: this method call should not return through the entire lifetime of this
+	// corruptible conduit factory.
+	// This is a client streaming gRPC implementation, and the input stream's lifecycle
+	// is tightly coupled with the lifecycle of this function call.
+	// Once it returns, the client stream is closed forever.
+	// Hence, we block the call and wait till a component shutdown.
```

Comment:
can we add a comment to the top of the function definition that it needs to be called in a goroutine because it's a blocking call? Otherwise, it's not clear that's how it needs to be used.

Or is it better to have this function start the goroutine here so clients don't have to do that?

Reply:
Added comment here: https://github.com/onflow/flow-go/pull/2807/commits/0fec3510d37b49e925a061dd2dc8ec09c1ae74e4